cnap/core/infereng.py

```diff
@@ -0,0 +1,332 @@
+        """
+        self._queue_topic = new_queue_topic
+
+    def __iter__(self) -> Iterator[Tuple[str, Any]]:
+        yield 'id', self.id
+        yield 'device', self.device
+        yield 'model_id', self.model_id
+
+class InferEngineManager:
+    """
+    To manage inference engines by registering them and searching for available engines.
+    """
+
+    _instance = None
+
+    def __init__(self, db: RuntimeDatabaseBase) -> None:
+        """
+        Initialize the InferEngineManager
+
+        Args:
+            db: RuntimeDatabaseBase object
+
+        Returns:
+            None
+        """
+        self._db = db
+
+    def search_engine(self, framework: str, target: str, device: str,
+                      model_name: str, model_version: str) -> Optional[InferenceInfo]:
+        """
+        Search the engine according to the framework and target requested.
+        return the InferenceInfo include the UUID of infer engine
+
+        Args:
+            framework: Framework of the model
+            target: Target of the model
+            device: Device type for infer engine
+            model_name: Name of the model
+            model_version: Version of the model
+
+        Returns:
+            InferenceInfo object if engine is found, None otherwise
+        """
+        # Get all available engines from database
+        engine_dicts = self._db.get_all_table_objects_dict(INFER_ENGINE_TABLE)
+
+        # Find the first engine that matches the desired framework and target
+        matching_engine_id, matching_engine_dict = next(
+            ((key, engine_dict) for key, engine_dict in engine_dicts.items()
+            if engine_dict['model']['details']['framework'] == framework
+            and engine_dict['model']['details']['target'] == target
+            and engine_dict['model']['details']['name'] == model_name
+            and engine_dict['model']['details']['version'] == model_version
+            and engine_dict['infer']['device'] == device
+            ),
+            (None, None))
+
+        if matching_engine_dict is None or matching_engine_id is None:
+            LOG.error("No matching engine found for framework for the request")
+            return None
+
+        # Convert matching engine dict to InferenceInfo object and return
+        matching_engine_info = InferenceInfo(
+            matching_engine_dict['infer']['device'],
+            matching_engine_dict['model']['id'])
+        matching_engine_info.id = matching_engine_id
+
+        return matching_engine_info
+
+    def register_engine(self, infer_info: InferenceInfo, model_info: ModelInfo) -> None:
+        """
+        Register the infer engine to database
+
+        Args:
+            infer_info: InferenceInfo object
+            model_info: ModelInfo object
+
+        Returns:
+            None
+        """
+        data = {
+        "infer": dict(infer_info),
+        "model": dict(model_info)
+        }
+
+        self._db.save_table_object_dict(
```

Comment:
Should we catch the error raised by `rtdb` here?

Reply:
fixed.